samza-core/src/main/java/org/apache/samza/clustermanager/ContainerManager.java

```diff
@@ -371,39 +397,62 @@ private boolean hasActiveContainerPlacementAction(String processorId) {
     return Optional.ofNullable(this.actions.get(processorId));
   }
 
+  private void updateContainerPlacementActionStatus(ContainerPlacementMetadata metadata,
+      ContainerPlacementMessage.StatusCode statusCode, String responseMessage) {
+    metadata.setActionStatus(statusCode, responseMessage);
+    writeContainerPlacementResponseMessage(metadata.getRequestMessage(), statusCode, responseMessage);
+    LOG.info("Status updated for ContainerPlacement action: {}", metadata);
+  }
+
+  private void writeContainerPlacementResponseMessage(ContainerPlacementRequestMessage requestMessage,
+      ContainerPlacementMessage.StatusCode statusCode, String responseMessage) {
+    containerPlacementMetadataStore.writeContainerPlacementResponseMessage(
+        ContainerPlacementResponseMessage.fromContainerPlacementRequestMessage(requestMessage, statusCode,
+            responseMessage, System.currentTimeMillis()));
+  }
+
   /**
-   * A valid container placement action is only issued for a running processor with a valid processor id which has no
-   * in flight container requests. Duplicate actions are handled by deduping on uuid
+   * If there is an existing inflight request or container is pending a start, the container placement action shall wait
+   * until this in-flight action is complete
```

Comment:
Just making a note here: This can be used in the future to allow adding ContainerPlacementRequestMessage that are scheduled to run in the future, e.g., when they are formulated and ordered by a KafkaCruiseControl-like controller-system.